File: app/javascript/dashboard/constants/subscriptionStatuses.js
// Subscription status constants matching backend Billing::SubscriptionStatuses
// These should be kept in sync with lib/billing/subscription_statuses.rb

// Stripe subscription statuses
export const SUBSCRIPTION_STATUSES = {
  ACTIVE: 'active',
  CANCELED: 'canceled',
  INCOMPLETE: 'incomplete',
  INCOMPLETE_EXPIRED: 'incomplete_expired',
  PAST_DUE: 'past_due',
  TRIALING: 'trialing',
  UNPAID: 'unpaid',
  PAUSED: 'paused',

  // Application-specific statuses
  INACTIVE: 'inactive',
};

// Status groups for easier filtering
export const PAID_STATUSES = [
  SUBSCRIPTION_STATUSES.ACTIVE,
  SUBSCRIPTION_STATUSES.TRIALING,
];

export const FAILED_PAYMENT_STATUSES = [
  SUBSCRIPTION_STATUSES.PAST_DUE,
  SUBSCRIPTION_STATUSES.CANCELED,
  SUBSCRIPTION_STATUSES.UNPAID,
];

// Helper functions
export const isPaidStatus = status => PAID_STATUSES.includes(status);
export const isFailedPaymentStatus = status =>
  FAILED_PAYMENT_STATUSES.includes(status);
export const shouldShowBanner = status =>
  [SUBSCRIPTION_STATUSES.PAST_DUE, SUBSCRIPTION_STATUSES.INACTIVE].includes(
    status
  );

File: lib/billing/subscription_statuses.rb
# frozen_string_literal: true

module Billing
  # Stripe subscription status constants
  # https://stripe.com/docs/api/subscriptions/object#subscription_object-status
  module SubscriptionStatuses
    # Stripe subscription statuses
    ACTIVE = 'active'           # Subscription is active and current
    CANCELED = 'canceled'       # Subscription has been canceled
    INCOMPLETE = 'incomplete'   # Initial payment not yet successful
    INCOMPLETE_EXPIRED = 'incomplete_expired' # Initial payment failed after retries
    PAST_DUE = 'past_due'      # Payment failed and subscription is past due
    TRIALING = 'trialing'      # Subscription is in trial period
    UNPAID = 'unpaid'          # Payment failed and subscription is unpaid
    PAUSED = 'paused'          # Subscription is paused

    # Application-specific statuses
    INACTIVE = 'inactive'       # Application status for ended subscriptions

    # Status groups for easier filtering
    PAID_STATUSES = [ACTIVE, TRIALING].freeze
    FAILED_PAYMENT_STATUSES = [PAST_DUE, CANCELED, UNPAID].freeze
    ALL_STRIPE_STATUSES = [
      ACTIVE, CANCELED, INCOMPLETE, INCOMPLETE_EXPIRED,
      PAST_DUE, TRIALING, UNPAID, PAUSED
    ].freeze

    # Check if status indicates subscription should have paid features
    def self.paid_status?(status)
      PAID_STATUSES.include?(status)
    end

    # Check if status indicates payment failure requiring community plan transition
    def self.failed_payment_status?(status)
      FAILED_PAYMENT_STATUSES.include?(status)
    end

    # Check if status is a valid Stripe status
    def self.valid_stripe_status?(status)
      ALL_STRIPE_STATUSES.include?(status)
    end
  end
end